modules/flink/src/main/java/org/apache/ignite/source/flink/IgniteSource.java

```diff
@@ -0,0 +1,203 @@
+
+    /** Cache name. */
+    private String cacheName;
+
+    /**
+     * Sets Ignite instance.
+     *
+     * @param ignite Ignite instance.
+     */
+    public void setIgnite(Ignite ignite) {
+        this.ignite = ignite;
+    }
+
+    /**
+     * Sets Event Batch Size.
+     *
+     * @param evtBatchSize Event Batch Size.
+     */
+    public void setEvtBatchSize(int evtBatchSize) {
+        this.evtBatchSize = evtBatchSize;
+    }
+
+    /**
+     * Sets Event Buffer timeout.
+     *
+     * @param evtBufTimeout Event Buffer timeout.
+     */
+    public void setEvtBufTimeout(int evtBufTimeout) {
+        this.evtBufTimeout = evtBufTimeout;
+    }
+
+    /** Gets Local Task Listener */
+    public TaskLocalListener getLocLsnr() { return this.locLsnr; }
+
+    /**
+     * Default IgniteSource constructor.
+     *
+     * @param cacheName Cache name.
+     */
+    public IgniteSource(String cacheName) { this.cacheName = cacheName; }
+
+    /**
+     * Starts Ignite source.
+     * @param filter User defined filter.
+     * @param cacheEvts Converts comma-delimited cache events strings to Ignite internal representation.
+     *
+     * @throws IgniteException If failed.
+     */
+    @SuppressWarnings("unchecked")
+    public void start(IgnitePredicate<CacheEvent> filter, int... cacheEvts) throws Exception {
+        A.notNull(cacheName, "Cache name");
+
+        stopped = false;
+
+        TaskRemoteFilter rmtLsnr = new TaskRemoteFilter(cacheName, filter);
+
+        try {
+            rmtLsnrId = ignite.events(ignite.cluster().forCacheNodes(cacheName))
+                 .remoteListen(locLsnr, rmtLsnr, cacheEvts);
+        }
+        catch (IgniteException e) {
+            log.error("Failed to register event listener!", e);
+
+            throw e;
+        }
+    }
+
+    /**
+     * Transfers data from grid.
+     *
+     * @param ctx SourceContext.
+     */
+    @Override public void run(SourceContext<CacheEvent> ctx) {
+        List<CacheEvent> evts = new ArrayList<>(evtBatchSize);
+
+        if (stopped)
+            return;
+
+        try{
+            while (!stopped) {
+                if (evtBuf.drainTo(evts, evtBatchSize) > 0) {
+                    synchronized (ctx.getCheckpointLock()) {
+                        for (CacheEvent evt : evts)
+                            ctx.collect(evt);
+
+                        evts = new ArrayList<>(evtBatchSize);
+                    }
+                }
+            }
+        } catch (Exception e){
+            log.error("Error while processing cache event of " + cacheName, e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
+        if (stopped)
+            return;
+
+        stopped = true;
+
+        if (rmtLsnrId != null && ignite != null) {
+            ignite.events(ignite.cluster().forCacheNodes(cacheName))
+                .stopRemoteListen(rmtLsnrId);
+
+            rmtLsnrId = null;
+
+            ignite.cache(cacheName).close();
```

Comment:
I'm not sure it is correct as IgniteSource uses a cache that was created outside of IgniteSource instance and can be used by someone else.